zingo-sync/src/scan.rs

```diff
@@ -0,0 +1,322 @@
+        Vec<(OutputId, orchard::Note, KeyId)>,
+    ),
+    (),
+>
+where
+    P: Parameters + Send + 'static,
+{
+    let mut runners = BatchRunners::<_, (), ()>::for_keys(100, scanning_keys);
+    for block in compact_blocks {
+        runners.add_block(parameters, block.clone()).unwrap();
+    }
+    runners.flush();
+
+    let mut incoming_sapling_outputs: Vec<(OutputId, sapling_crypto::Note, KeyId)> = Vec::new();
+    let mut incoming_orchard_outputs: Vec<(OutputId, orchard::Note, KeyId)> = Vec::new();
+
+    // TODO: add outgoing decryption
+    for block in compact_blocks {
+        for transaction in block.vtx.iter() {
+            let decrypted_sapling_outputs = runners
+                .sapling
+                .collect_results(block.hash(), transaction.txid());
+            decrypted_sapling_outputs
+                .iter()
+                .for_each(|((txid, output_index), output)| {
+                    incoming_sapling_outputs.push((
+                        OutputId::from_parts(*txid, *output_index),
+                        output.note.clone(),
+                        output.ivk_tag,
+                    ));
+                });
+
+            let decrypted_orchard_outputs = runners
+                .orchard
+                .collect_results(block.hash(), transaction.txid());
+            decrypted_orchard_outputs
+                .iter()
+                .for_each(|((txid, output_index), output)| {
+                    incoming_orchard_outputs.push((
+                        OutputId::from_parts(*txid, *output_index),
+                        output.note,
+                        output.ivk_tag,
+                    ));
+                });
+        }
+    }
+
+    Ok((incoming_sapling_outputs, incoming_orchard_outputs))
+}
+
+// checks height and hash continuity of a batch of compact blocks.
+// takes the last wallet compact block of the adjacent lower scan range, if available.
+fn check_continuity(
+    compact_blocks: &[CompactBlock],
+    previous_compact_block: Option<&WalletCompactBlock>,
+) -> Result<(), ()> {
+    let mut prev_height: Option<BlockHeight> = None;
+    let mut prev_hash: Option<BlockHash> = None;
+
+    if let Some(prev) = previous_compact_block {
+        prev_height = Some(prev.block_height());
+        prev_hash = Some(prev.block_hash());
+    }
+
+    for block in compact_blocks {
+        if let Some(prev_height) = prev_height {
+            if block.height() != prev_height + 1 {
+                panic!("height discontinuity");
+            }
+        }
+
+        if let Some(prev_hash) = prev_hash {
+            if block.prev_hash() != prev_hash {
+                panic!("hash discontinuity");
+            }
+        }
+
+        prev_height = Some(block.height());
+        prev_hash = Some(block.hash());
+    }
+
+    Ok(())
+}
+
+// calculates nullifiers and positions for a given compact transaction
+fn calculate_nullifiers_and_positions<D, K, Nf>(
+    transaction: &CompactTx,
```

Comment:
putting the CompactTx parameter is mostly redundant

Reply:
yeah i agree this isnt optimal in this PR, i have fixed this in a following PR where we dont need to filter on txid